lib/DataLayoutAnalysis/Backend/DLAMakeModelTypes.cpp

```diff
@@ -29,381 +29,301 @@ static Logger<> Log("dla-make-model");
+    rc_return FieldType;
+
+  // If the stride is larger than the size of the inner element, we need to
+  // reserve space after each element, hence we wrap it in a struct.
+  if (Stride > FieldSize) {
+    revng_log(Log, "Creating wrapper");
+    // Create a wrapper to hold each element (except for the last one)
+    // together with its trailing padding
+    UpcastableType ElemWrapper = makeStructWrapper(FieldType.copy(),
+                                                   Model,
+                                                   /*offset*/ 0,
+                                                   /*size*/ Stride);
+
+    // Now make this an array.
+    UpcastableType WrappedArray = model::ArrayType::make(std::move(ElemWrapper),
+                                                         (NumElements - 1));
+
+    const uint64_t LastElemOffset = Stride * (NumElements - 1);
+    const uint64_t ArrayWrapperSize = LastElemOffset + FieldSize;
+
+    // Create a wrapper to hold the array + the last element, which does not
+    // need trailing padding.
+    UpcastableType ArrayWrapper = makeStructWrapper(std::move(WrappedArray),
+                                                    Model,
+                                                    /*offset*/ 0,
+                                                    /*size*/ ArrayWrapperSize);
+
+    // Insert the last element
+    model::StructDefinition &Struct = ArrayWrapper->toStruct();
+
+    // Insert the rest of the array
+    Struct.addField(LastElemOffset, std::move(FieldType));
+    Struct.Size() = ArrayWrapperSize;
+
+    FieldType = std::move(ArrayWrapper);
+  } else {
+    revng_log(Log, "Wrapping it into an array with no struct wrappers");
+    FieldType = model::ArrayType::make(std::move(FieldType), NumElements);
+  }
+
+  rc_return FieldType;
 }
 
-/// Create the right type for a pointer node by following pointer
-/// edges until you find a pointee.
-static std::pair<size_t, const LTSN *>
-getNumPointersAndPointee(const LTSN *Node) {
-  revng_assert(isPointerNode(Node));
-  const LTSN *PointerNode = Node;
-  size_t NumPointers = 0;
-  while (isPointerNode(PointerNode)) {
-    revng_log(Log, "Following Pointer " << PointerNode->ID);
-    ++NumPointers;
-    PointerNode = getPointeeNode(PointerNode);
+enum class FieldKind {
+  StructField,
+  UnionField
+};
+
+/// Generate a field's type from an offset expression, generating also
+/// intermediate array and pointer types if needed.
+template<FieldKind K>
+static UpcastableType makeFieldType(const LTSN *FieldNode,
+                                    const OffsetExpression &OE,
+                                    NodeToUpcastableTypeMap &NodesToTypes,
+                                    TupleTree<model::Binary> &Model) {
+  if (Log.isEnabled()) {
+    std::string OEStr;
+    {
+      llvm::raw_string_ostream OS{ OEStr };
+      OE.print(OS);
+    }
+    revng_log(Log, "makeFieldType: FieldNode ID: " << FieldNode->ID << OEStr);
```

Comment:
Not having a space and/or a new line between the ID and the `OffsetExpression` makes me suspicious.

Unless `OffsetExpression` includes it for some reason. In which case, sorry for the noise.

Reply:
Nah it's ok. The logs stay more compact this way. And printing `OffsetExpression` doesn't have a newline but it has some leading text for formatting, so the output is still pretty readable.

Reply:
Gotcha, sorry for the noise then.